lib/r10k/source/exec.rb

```diff
@@ -0,0 +1,40 @@
+require 'open3'
+require 'json'
+require 'yaml'
+
+class R10K::Source::Exec < R10K::Source::Hash
+  R10K::Source.register(:exec, self)
+
+  def initialize(name, basedir, options = {})
+    unless command = options[:command]
+      raise ConfigError, _('Environment source %{name} missing required parameter: command') % {name: name}
+    end
+
+    begin
+      stdout, stderr, status = Open3.capture3(command)
+    rescue Errno::EACCES, Errno::ENOENT => e
+      raise R10K::Error, _("Error executing command for exec source %{name}: %{msg}") % {name: name, msg: e.message}
+    end
+
+    unless status.success?
+      raise R10K::Error, _("Error executing exec command for source %{name}: %{cmd}. Command exited %{code}\n" \
+                           "stdout:\n%{out}\n" \
+                           "stderr:\n%{err}") % {name: name, cmd: command, code: status.exitstaus.to_s, out: stdout, err: stderr}
+    end
+
+    begin
+      environments = JSON.parse(stdout) rescue YAML.load(stdout)
+    rescue Psych::SyntaxError => e
+      raise R10K::Error, _('Error parsing command output for exec source %{name}; ' \
+                           "tried JSON first, then YAML; YAML parsing failed with: %{msg}\n" \
+                           "stdout:\n%{out}") % {name: name, msg: e.message, out: stdout}
```

Comment:
any reason why JSON is tried first? would there be any reason we'd want that error message too?

Reply:
would it be worth giving a way to specify which format to expect?

Reply:
Yeah, this could get cleaned up to expose the JSON parsing error too if all fails. Mostly just done this way the first time around for code brevity.

JSON is tried first because it will correctly read valid JSON, and will fail on anything else. YAML goes second because it will actually parse JSON as valid input _most of the time_, but has edge cases that could produce something weird. Trying to parse as JSON first avoids that.

Design-wise, we should not let the user specify. We can figure it out easily so there's no benefit to making the user think about it or care during configuration.

Reply:
Modified in new commit to show parsing error results of JSON and YAML attempts, if both fail.

Reply:
ahh okay, that makes sense to me. Thanks!